md/documentation-template.j2

```diff
@@ -0,0 +1,1273 @@
+### {{ protocol_message_offset }}.1. Custom Data Types Used In The Protocol
+{% for definition in custom_definitions %}
+{% set custom_types = definition.get('customTypes', None) %}
+{% if custom_types is not none %}
+{% for custom_type in custom_types %}
+#### {{ protocol_message_offset }}.1.{{ loop.index }}. {{ custom_type.name }}
+**Parameters**
+
+| Name | Type | Nullable | Available Since |
+| ---- | ---- | -------- | --------------- |
+            {% for param in custom_type.params %}
+| {{ param.name }} | {{ convert_type(param.type) }} | {{ param.nullable }} | {{ param.since }} |
+            {% endfor %}
+
+{% endfor %}
+{% endif %}
+{% endfor %}
+{% for service in services %}
+    {% set service_id = service.id + 2 %}
+
+### {{ protocol_message_offset }}.{{ service_id }}. {{ service.name }}
+    {% for method in service.methods %}
+
+#### {{ protocol_message_offset }}.{{ service_id }}.{{ method.id }}. {{ service.name }}.{{ method.name|capital }}
+```
+{{ method.doc }}
+```
+
+**Available since:** {{ method.since }}
+
+#### Request Message
+**Message Type:** {{ "0x%02x%02x%02x"|format(service.id, method.id, 0) }}
+        {% if method.request.params|length > 0 %}
+
+**Partition Identifier:** {{ resolve_partition_identifier(method.request.partitionIdentifier) }}
+
+| Name | Type | Nullable | Description | Available Since |
+| ---- | ---- | -------- | ----------- | --------------- |
+            {% for param in method.request.params %}
+| {{ param.name }} | {{ convert_type(param.type) }} | {{ param.nullable }} | {{ param.doc.splitlines()|join(" ") }} | {{ param.since }} |
+            {% endfor %}
+        {% else %}
+
+Header only request message, no message body exist.
+        {% endif %}
+
+#### Response Message
+**Message Type:** {{ "0x%02x%02x%02x"|format(service.id, method.id, 1) }}
+        {% if method.response.params|length > 0 %}
+
+| Name | Type | Nullable | Description | Available Since |
+| ---- | ---- | -------- | ----------- | --------------- |
+            {% for param in method.response.params %}
+| {{ param.name }} | {{ convert_type(param.type) }} | {{ param.nullable }} | {{ param.doc.splitlines()|join(" ") }} | {{ param.since }} |
+            {% endfor %}
+        {% else %}
+
+Header only response message, no message body exist.
+        {% endif %}
+        {% if method.events|length > 0 %}
+
+#### Event Message
+            {% for event in method.events %}
+
+##### {{ event.name }}
+**Message Type:** {{ "0x%02x%02x%02x"|format(service.id, method.id, 2 + loop.index) }}
+
+| Name | Type | Nullable | Description | Available Since |
+| ---- | ---- | -------- | ----------- | --------------- |
+            {% for param in event.params %}
+| {{ param.name }} | {{ convert_type(param.type) }} | {{ param.nullable }} | {{ param.doc.splitlines()|join(" ") }} | {{ param.since }} |
+            {% endfor %}
+            {% endfor %}
+        {% endif %}
+    {% endfor %}
+{% endfor %}
```

Comment:
nit: do we to place a copyright/trademark section somewhere in the generated document?

Reply:
added a seperate section for it